// @eth-optimism/hardhat-ovm/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const node_fetch_1 = __importDefault(require("node-fetch"));
const config_1 = require("hardhat/config");
const global_dir_1 = require("hardhat/internal/util/global-dir");
const artifacts_1 = require("hardhat/internal/artifacts");
const task_names_1 = require("hardhat/builtin-tasks/task-names");
require("./type-extensions");
const OPTIMISM_SOLC_VERSION_URL = 'https://api.github.com/repos/ethereum-optimism/solc-bin/git/refs/heads/gh-pages';
const OPTIMISM_SOLC_BIN_URL = 'https://raw.githubusercontent.com/ethereum-optimism/solc-bin/gh-pages/bin';
const DEFAULT_OVM_SOLC_VERSION = '0.7.6';
const getOvmSolcPath = async (version) => {
    if (process.env.__DANGEROUS_OVM_IGNORE_ERRORS__) {
        console.log('\n\n__DANGEROUS_OVM_IGNORE_ERRORS__ IS ENABLED!\n\n');
        version += '-no_errors';
    }
    const ovmCompilersCache = path.join(await global_dir_1.getCompilersDir(), 'ovm');
    if (!fs.existsSync(ovmCompilersCache))
        [fs.mkdirSync(ovmCompilersCache, { recursive: true })];
    const remoteCompilerVersion = await (await node_fetch_1.default(OPTIMISM_SOLC_VERSION_URL)).text();
    let cachedCompilerVersion = '';
    const cachedCompilerVersionPath = path.join(ovmCompilersCache, `version-info-${version}.json`);
    if (fs.existsSync(cachedCompilerVersionPath)) {
        cachedCompilerVersion = fs
            .readFileSync(cachedCompilerVersionPath)
            .toString();
    }
    const cachedCompilerPath = path.join(ovmCompilersCache, `${version}.js`);
    if (remoteCompilerVersion === cachedCompilerVersion &&
        fs.existsSync(cachedCompilerPath)) {
        return cachedCompilerPath;
    }
    console.log(`Downloading OVM compiler version ${version}`);
    const compilerContentResponse = await node_fetch_1.default(OPTIMISM_SOLC_BIN_URL + `/soljson-v${version}.js`);
    if (!compilerContentResponse.ok) {
        throw new Error(`Unable to download OVM compiler version ${version}. Are you sure that version exists?`);
    }
    const compilerContent = await compilerContentResponse.text();
    fs.writeFileSync(cachedCompilerPath, compilerContent);
    fs.writeFileSync(cachedCompilerVersionPath, remoteCompilerVersion);
    return cachedCompilerPath;
};
config_1.subtask(task_names_1.TASK_COMPILE_SOLIDITY_RUN_SOLC, async (args, hre, runSuper) => {
    var _a, _b;
    const ignoreRxList = hre.network.config.ignoreRxList || [];
    const ignore = (filename) => ignoreRxList.reduce((ignored, rx) => ignored || new RegExp(rx).test(filename), false);
    if (hre.network.ovm !== true) {
        for (const file of Object.keys(args.input.sources)) {
            if (args.input.sources[file].content.includes('// @unsupported: evm') ||
                ignore(file)) {
                delete args.input.sources[file];
            }
            else {
            }
        }
        return runSuper(args);
    }
    let ovmSolcVersion = DEFAULT_OVM_SOLC_VERSION;
    if ((_b = (_a = hre.config) === null || _a === void 0 ? void 0 : _a.ovm) === null || _b === void 0 ? void 0 : _b.solcVersion) {
        ovmSolcVersion = hre.config.ovm.solcVersion;
    }
    const ovmSolcPath = await getOvmSolcPath(ovmSolcVersion);
    const ovmInput = {
        language: 'Solidity',
        sources: {},
        settings: args.input.settings,
    };
    for (const file of Object.keys(args.input.sources)) {
        if (!args.input.sources[file].content.includes('// @unsupported: ovm') &&
            !ignore(file)) {
            ovmInput.sources[file] = args.input.sources[file];
        }
    }
    const ovmOutput = await hre.run(task_names_1.TASK_COMPILE_SOLIDITY_RUN_SOLCJS, {
        input: ovmInput,
        solcJsPath: ovmSolcPath,
    });
    ovmOutput.errors = (ovmOutput.errors || []).map((error) => {
        if (error.severity === 'error') {
            error.formattedMessage = `OVM Compiler Error (insert "// @unsupported: ovm" if you don't want this file to be compiled for the OVM):\n ${error.formattedMessage}`;
        }
        return error;
    });
    return ovmOutput;
});
config_1.extendEnvironment((hre) => {
    if (hre.network.config.ovm) {
        hre.network.ovm = hre.network.config.ovm;
        let artifactsPath = hre.config.paths.artifacts;
        if (!artifactsPath.endsWith('-ovm')) {
            artifactsPath = artifactsPath + '-ovm';
        }
        let cachePath = hre.config.paths.cache;
        if (!cachePath.endsWith('-ovm')) {
            cachePath = cachePath + '-ovm';
        }
        hre.config.paths.artifacts = artifactsPath;
        hre.config.paths.cache = cachePath;
        hre.artifacts = new artifacts_1.Artifacts(artifactsPath);
    }
});
